homeassistant/components/homekit/type_fans.py

```diff
@@ -83,6 +98,56 @@ def set_oscillating(self, value):
                   ATTR_OSCILLATING: oscillating}
         self.call_service(DOMAIN, SERVICE_OSCILLATE, params, oscillating)
 
+    def set_speed_list(self, speed_list):
+        """Set and validate the speed list from the fan."""
+        if self.speed_list == speed_list:
+            return
+        if not speed_list:
+            self.speed_list = []
+            return
+        if SPEED_OFF != speed_list[0]:
+            _LOGGER.warning(
+                "%s: %s (%s) does not contain the speed setting "
+                "%s as its first element. "
+                "Assuming that %s is equivalent to 'off'.""",
+                self.entity_id, ATTR_SPEED_LIST, speed_list,
+                SPEED_OFF, speed_list[0])
+        self.speed_list = speed_list
+
+    def speed_from_percentage(self, percentage):
+        """Convert a value from 0-100 to a value in the speed list."""
+        if not self.speed_list:
+            return None
+        if percentage <= 0:
+            return self.speed_list[0]
+        if percentage >= 100:
+            return self.speed_list[-1]
+        index = int(percentage / 100 * len(self.speed_list))
+        return self.speed_list[index]
+
+    def percentage_from_speed(self, speed):
+        """Convert a value in the speed list into a value from 0-100."""
+        try:
+            index = self.speed_list.index(speed)
+            # By dividing by len(speed_list) -1 the following
+            # desired attributes hold true:
+            # * index = 0 => 0%, equal to "off"
+            # * index = len(speed_list) - 1 => 100 %
+            # * all other indices are equally distributed
+            return index * 100 / (len(self.speed_list) - 1)
+        except ValueError:
+            # speed_list does not contain the speed
+            return None
+
+    def set_speed(self, value):
+        """Set state if call came from HomeKit."""
+        _LOGGER.debug('%s: Set speed to %d', self.entity_id, value)
+        self._flag[CHAR_ROTATION_SPEED] = True
+        speed = self.speed_from_percentage(value)
+        params = {ATTR_ENTITY_ID: self.entity_id,
+                  ATTR_SPEED: speed}
```

Comment:
```py
ATTR_SPEED: speed_to_states(speed, self.speed_list_type)

Reply:
Now using generic `self.speed_mapping.speed_to_states(value)` to map the attribute to a percentage from HA states